common/src/main/java/com/graphicsfuzz/common/util/GloballyTruncateLoops.java

```diff
@@ -39,9 +64,90 @@
    */
   public static void truncate(ShaderJob shaderJob, int loopLimit, String loopCountName,
                               String loopBoundName) {
-    // TODO(https://github.com/google/graphicsfuzz/issues/866): Implement loop limiting
-    //  logic.
-    throw new RuntimeException("Not implemented yet.");
-  }
+    for (TranslationUnit tu : shaderJob.getShaders()) {
+      Declaration firstNonPrecisionDeclaration = null;
+      for (Declaration decl : tu.getTopLevelDeclarations()) {
+        if (decl instanceof PrecisionDeclaration) {
+          continue;
+        }
+        firstNonPrecisionDeclaration = decl;
+        break;
+      }
+      // Add loop bound variable
+      tu.addDeclarationBefore(new VariablesDeclaration(new QualifiedType(BasicType.INT,
+              Arrays.asList(TypeQualifier.CONST)),new VariableDeclInfo(loopBoundName,null,
+              new Initializer(new IntConstantExpr(new Integer(loopLimit).toString())))),
+          firstNonPrecisionDeclaration);
+      // Add loop count variable
+      tu.addDeclarationBefore(new VariablesDeclaration(BasicType.INT,
+              new VariableDeclInfo(loopCountName,null,
+              new Initializer(new IntConstantExpr("0")))),
+          firstNonPrecisionDeclaration);
+      // Traverse tree for the rest of the transformations
+      new StandardVisitor() {
+        // Converts loop condition from "x" to "(x) && (lc<lb)"
+        Expr buildCondition(Expr originalCondition) {
+          return new BinaryExpr(
+              new ParenExpr(originalCondition),
+              new ParenExpr(
+                  new BinaryExpr(new VariableIdentifierExpr(loopCountName),
+                      new VariableIdentifierExpr(loopBoundName),
+                      BinOp.LT)),
+              BinOp.LAND);
+        }
+
+        // Check if the loop condition is just "false"
+        boolean isTrueLoop(Expr conditionExpr) {
+          return !(conditionExpr instanceof BoolConstantExpr
+              && !((BoolConstantExpr)conditionExpr).getIsTrue());
+        }
+
+        @Override
+        public void visitForStmt(ForStmt forStmt) {
+          if (isTrueLoop(forStmt.getCondition())) {
+            forStmt.setCondition(buildCondition(forStmt.getCondition()));
+            // Add block statement if it's missing
+            if (!(forStmt.getBody() instanceof BlockStmt)) {
+              forStmt.setBody(new BlockStmt(Arrays.asList(forStmt.getBody()), true));
+            }
+            // Add loop count increment a start of body block
+            ((BlockStmt) forStmt.getBody()).insertStmt(0,
+                new ExprStmt(new UnaryExpr(new VariableIdentifierExpr(loopCountName),
+                    UnOp.POST_INC)));
+          }
```

Comment:
Can you introduce:

private void handleLoopStmt(LoopStmt loopStmt) {
  // Do the condition check, replacement, and new statement adding
}

Because the same work is required for all three kinds of loop (including do..while, I think, as per my comment below), you could then do the "for" case as folows:

public void visitForStmt(ForStmt forStmt) {
  handleLoopStmt(forStmt);
  super.visitForStmt(forStmt);
}

and similar for the other cases.

Reply:
All three loop types were handled in a slightly different way - do-while always has body with a block statement, but it won't hurt to check for it. while... has a diffent scope rule, but that can be handled via a simple parameter, which I added.